google-cloud-firestore/lib/google/cloud/firestore/query_listener.rb

```diff
@@ -0,0 +1,566 @@
+                added_doc = removed_doc
+              end
+
+              add_doc_to_tree added_doc
+            end
+
+            @resume_token = resume_token
+            @read_time = read_time
+            clear_pending
+          end
+
+          def changes?
+            # Act like there are changes if we have never run before
+            return true if @old_order.nil?
+            added_paths, deleted_paths, changed_paths = \
+              change_paths current_order, @old_order
+            added_paths.any? || deleted_paths.any? || changed_paths.any?
+          end
+
+          def current_docs
+            @tree.keys
+          end
+
+          def order_for docs
+            Hash[docs.map { |doc| [doc.path, doc.updated_at] }]
+          end
+
+          def current_order
+            order_for current_docs
+          end
+
+          def build_query_snapshot
+            # If this is the first time building, set to empty hash
+            @old_order ||= {}
+
+            # Get the new set of documents, changes, order
+            docs = current_docs
+            new_order = order_for docs
+            changes = build_changes new_order, @old_order
+            @old_order = new_order
+
+            QuerySnapshot.from_docs @query, docs, changes, @read_time
+          end
+
+          protected
+
+          def query_comparison_proc a, b
+            last_direction = nil
+
+            @query.query.order_by.each do |order|
+              field_path = order.field.field_path # "__name__"
+              last_direction = order.direction
+
+              comp = compare_fields field_path, a[field_path], b[field_path]
+              comp = apply_direction comp, last_direction
+              return comp unless comp.zero?
+            end
+
+            apply_direction compare_paths(a.path, b.path), last_direction
+          end
+
+          def apply_direction comparision, direction
+            return 0 - comparision if direction == :DESCENDING
+
+            comparision
+          end
+
+          def compare_fields field_path, a_value, b_value
+            if field_path == "__name__".freeze
+              return compare_paths a_value, b_value
+            end
+
+            compare_values a_value, b_value
+          end
+
+          def compare_values a_value, b_value
+            field_comparison(a_value) <=> field_comparison(b_value)
+          end
+
+          def compare_paths a_path, b_path
+            path_comparison(a_path) <=> path_comparison(b_path)
+          end
+
+          def field_comparison value
+            [field_type(value), field_value(value)]
+          end
+
+          def path_comparison value
+            value.split("/")
+          end
+
+          def field_type value
+            return 0 if value.nil?
+            return 1 if value == false
+            return 1 if value == true
+            return 2 if value.respond_to?(:nan?) && value.nan?
+            return 3 if value.is_a? Numeric
```

Comment:
Hm, this is a pretty clever handling of the NaN case, but I am still a little bit opposed. We don't have a sane way to enforce ordering across all SDKs and the backend if the code doesn't match.

Reply:
I assume this comment was intended for line 474, and not 475. Can you confirm that your concern is separating NaN values from every other numeric value?

If so, we are doing this because of the way Ruby's comparison works for NaN values. Like `nil`, they don't compare to each other.

```ruby
num1 = Float::NAN
num2 = Float::NAN
num1 == num2 #=> false
```

Worse, comparing a NaN value will return a `nil`, which is considered and error case:

```ruby
num1 <=> num2 #=> nil
```

So we separate them so all NaN values are ordered before all other numbers (or after if the order is descending), and we don't actually compare the NaN values, we compare 0 instead. I believe we are enforcing order correctly, and have tests demonstrating this.

Reply:
Yes, that was my concern (line 474).

Can you add a comment here that is divergence is by design and that it should not be copied to other clients? Something like

```
The backend ordering semantics treats NaN and Numbers as the same type, and then internally orders NaNs before Numbers. In Ruby, we can achieve the same by separating the types and thus, we intentionally diverge from type order that is used in the backend and in the other SDKs.
```

Reply:
Added.